File: medium/Linked List Cycle II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        if head is None:
            return None
        node_list = [head]
        while head.next is not None:
            if head.next in node_list:
                return head.next
            head = head.next
            node_list.append(head)
        return None
      
      
      
      
# # Definition for singly-linked list.
# # class ListNode:
# #     def __init__(self, x):
# #         self.val = x
# #         self.next = None

# class Solution:
#     def detectCycle(self, head: ListNode) -> ListNode:
#         if head is None:
#             return None
#         node_list = {head:1}
#         while head.next is not None:
#             if head.next in node_list:
#                 return head.next
#             head = head.next
#             node_list[head] = 1
#         return None

File: cpp/medium/Linked_List_Cycle_II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if (head == nullptr) return nullptr;
        unordered_set<ListNode*> node_set;
        node_set.insert(head);
        while (head->next != nullptr) {
            if (node_set.find(head->next) == node_set.end()) node_set.insert(head);
            else return head->next;
            head = head->next;
        }
        return nullptr;
    }
};
